bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/NonSiderealNameEditor.scala

```diff
@@ -0,0 +1,152 @@
+import edu.gemini.spModel.target.system.ConicTarget
+
+import jsky.app.ot.gemini.editor.targetComponent.TelescopePosEditor
+import jsky.util.gui.{DialogUtil, TextBoxWidget, TextBoxWidgetWatcher}
+
+
+import Horizons.HorizonsIO
+
+import scala.swing.Swing
+import scalaz._, Scalaz._, effect.IO
+
+final class NonSiderealNameEditor extends TelescopePosEditor with ReentrancyHack {
+
+  private[this] var site = Option.empty[Site]
+  private[this] var spt = new SPTarget // never null
+
+  def lookup(site: Option[Site]): Unit = {
+    import HorizonsService2.{ Search, Row }, Search._
+
+    /// Some IO actions for looking up targets and epherimides
+
+    val askSearch: HS2[Option[Search[_ <: HorizonsDesignation]]] =
+      HS2.delay {
+        case class Wrapper(unwrap: Search[_ <: HorizonsDesignation]) {
+          override def toString = unwrap.productPrefix
+        }
+        Option {
+          JOptionPane.showInputDialog(
+            name,
+            "Select the HORIZONS target type:",
+            "Horizons Search",
+            JOptionPane.QUESTION_MESSAGE,
+            null, // TODO: icon
+            List(Comet, Asteroid, MajorBody).map(f => Wrapper(f(name.getText))).toArray[Object],
+            null)
+        } .map(_.asInstanceOf[Wrapper].unwrap)
+      }
+
+    def show(msg: String): HS2[Unit] =
+      HS2.delay(Swing.onEDT(GlassLabel.show(SwingUtilities.getRootPane(name), msg)))
+
+    def hide: HS2[Unit] =
+      HS2.delay(Swing.onEDT(GlassLabel.hide(SwingUtilities.getRootPane(name))))
+
+    def lookup(d: HorizonsDesignation, site: Site): HS2[Ephemeris] =
+      show("Fetching Ephemeris...") *> HorizonsService2.lookupEphemeris(d, site, 1000) // arbitrary
+
+    def updateDesignation(hd: HorizonsDesignation, name: String): HS2[Unit] =
+      HS2.delay {
+        val t0 = Target.name.set(spt.getNewTarget, name)
+        Target.horizonsDesignation.set(t0, Some(hd)).foreach(spt.setNewTarget)
+      }
+
+    def updateEphem(e: Ephemeris): HS2[Unit] =
+      HS2.delay(Target.ephemeris.set(spt.getNewTarget, e).foreach(spt.setNewTarget))
+
+    def manyResults(rs: List[Row[_ <: HorizonsDesignation]]): HS2[Unit] =
+      HS2.delay {
+        case class Wrapper(unwrap: Row[_ <: HorizonsDesignation]) {
+          override def toString = unwrap.name + " - " + unwrap.a.des
+        }
+        Option {
+          JOptionPane.showInputDialog(
+            name,
+            "Multiple results were found. Please disambiguate:",
+            "Horizons Search",
+            JOptionPane.QUESTION_MESSAGE,
+            null, // TODO: icon
+            rs.map(Wrapper).sortBy(_.toString).toArray[Object],
+            null)
+        } .map(_.asInstanceOf[Wrapper].unwrap)
+      } >>= {
+        case Some(r) => oneResult(r)
+        case None    => ().point[HS2]
+      }
+
+    def oneResult(r: Row[_ <: HorizonsDesignation]): HS2[Unit] =
+      updateDesignation(r.a, r.name).as(site) >>= {
+        case Some(site) => lookup(r.a, site) <* hide >>= updateEphem
```

Comment:
It means lookup and then hide the glass pane, but we don't care about the value of `hide`. These operators are defined for all applicative functors:

- `a *> b` is the same as `(a |@| b)((aa, bb) => bb)`
- `a <* b` is the same as  `(a |@| b)((aa, bb) => aa)`

So the expression `lookup(r.a, site) <* hide >>= updateEphem` is just shorthand for

```scala
for {
  x <- lookup(r.a, site)
  _ <- hide
  e <- updateEphem(x)
} yield e
```